src/blog/2024/001-bpmn-js-type-definitions.md

```diff
@@ -0,0 +1,94 @@
+---
+
+title: Type Declarations for bpmn-js
+description: "Learn how to use the type declarations that ship with bpmn-js@13 to get the most out of your editor tooling."
+layout: blogpost.hbs
+slug: 2024-bpmn-js-type-declarations
+preview_image: http://bpmn.io/assets/attachments/blog/2024/001-code-completion.png
+url: https://bpmn.io/blog/posts/2024-bpmn-js-type-declarations.html
+author:
+- Nico Rehwaldt <https://github.com/nikku>
+published: 2024-01-09 12:00
+
+releases:
+  - 'bpmn-js@13.0.0'
+  - 'diagram-js@12.0.0'
+
+---
+
+<p class="introduction">
+  Earlier last year, [bpmn-js@13](https://github.com/bpmn-io/bpmn-js/blob/main/CHANGELOG.md#1300) shipped with a single feature: Built-in [type declarations](https://www.typescriptlang.org/docs/handbook/2/type-declarations.html). Learn how they help you get the most out of your editor tooling as you work with bpmn-js, whether through code completion, introspection, or validation. Look behind the scenes and understand how we ended up creating these declarations.
+</p>
+
+<!-- continue -->
+
+Advanced code intelligence for JavaScript (and TypeScript) in modern code editors relies on a single feature: [Type declarations](https://www.typescriptlang.org/docs/handbook/2/type-declarations.html). Type declarations, shipped with the libraries you use, superpower your code editor, allow for auto-completion, and enable static syntax checking.
+
+With [bpmn-js@13](https://github.com/bpmn-io/bpmn-js/blob/main/CHANGELOG.md#1300), we made this utility available for users, too:
+
+<div class="figure full-size">
+  <img src="{{ assets }}/attachments/blog/2024/001-code-completion.gif" alt="Code completion for bpmn-js in a modern editor.">
+
+  <p class="caption">
+    Code completion for bpmn-js in a modern editor.
+  </p>
+</div>
+
+## Using the Declarations
+
+Type declarations ship out of the box, starting with `bpmn-js@13`. A modern code editor picks those up automatically or with minimal configuration effort.
+
+We implemented types so that event listeners (`Diagram#on`) and service getters (`Diagram#get`) are generics. Cast them to the appropriate type as you use the library:
+
+```javascript
+import BpmnViewer from 'bpmn-js/lib/Viewer';
+
+/**
+ * @typedef {
+ *   import('diagram-js/lib/core/ElementRegistry').default
+ * } ElementRegistry
+ */
+
+const viewer = new BpmnViewer();
+
+/**
+ * @type { ElementRegistry }
+ */
+const elementRegistry = viewer.get('elementRegistry');
+
+const element = elementRegistry.get('MY_TASK');
+
+console.log(element.id); // MY_TASK
+```
+
+Check out the [full types example](https://github.com/bpmn-io/bpmn-js-examples/blob/main/types) to learn more about bpmn-js type declarations and their use cases.
```

Comment:
I assume we're going to create this example?

Reply:
Cf. https://github.com/bpmn-io/bpmn-js-examples/pull/268

Reply:
😅